src/test/ui/lint/lint-unnecessary-parens.rs

```diff
@@ -22,8 +22,8 @@ fn main() {
     match (true) { //~ ERROR unnecessary parentheses around `match` head expression
         _ => {}
     }
-    if let 1 = (1) {} //~ ERROR unnecessary parentheses around `if let` head expression
-    while let 1 = (2) {} //~ ERROR unnecessary parentheses around `while let` head expression
+    if let 1 = (1) {} //~ ERROR unnecessary parentheses around `let` scrutinee
```

Comment:
Not sure if we want to throw the term `scrutinee` at users. cc @rust-lang/wg-diagnostics

Reply:
This is standard terminology that all typed languages with pattern matching that I know of use:
https://github.com/rust-lang-nursery/reference/pull/496#issue-241587218. And the reference uses it.

Reply:
It's horrible terminology to be honest. I'd be freaked out as a newbie, seeing the term. I have a suspicion the diagnostics guys won't like it much either.

Reply:
I think introducing a new user-facing term requires a lot more oversight here. The reference is not user-facing, this is.

We should keep it as-is for now, perhaps introducing a new term if the docs team can settle on one.

Reply:
The terminology comes from "scrutinizing" the expression according to the patterns, which seems quite logical to me. I don't understand why a newbie should freak out when seeing this. Moreover, "scrutinee" actually has google results which makes sense -- "head expression" does not.

As a general note... I quite dislike when we try to invent custom terminology for concepts that already have well established terms of art.

Reply:
"head expression" makes sense in context, "scrutinee" doesn't if you haven't heard of it.

I have literally never heard this term before. "head expression" is a choice taken to _not_ invent new terminology, and instead compose _existing_ terminology into something that makes sense. This works for people who have and haven't used other languages with this term. Scrutinee only helps people who have heard this term before, and I don't think that's a super large group. Anecdotally I've learned quite a few PLs and never come across this.

Reply:
I don't know how "head" makes sense here; in `let p = e`, what does "head" mean? -- it's not the thing that comes first, the pattern is. For "scrutinee" you can at least google the term. If we append "expression" then it might be even clearer.

> Anecdotally I've learned quite a few PLs and never come across this.

[Haskell](https://downloads.haskell.org/~ghc/7.6.2/docs/html/users_guide/informal-semantics.html), [OCaml](https://caml.inria.fr/pub/docs/manual-ocaml/extn.html#sec252), [Idris](https://www.idris-lang.org/idris-0-9-11-released/), [Agda](https://media.readthedocs.org/pdf/agda/v2.5.3/agda.pdf), [Scala](https://issues.scala-lang.org/browse/SI-7171), and [ML](https://samth.github.io/match-ifl-full.pdf).

> We should keep it as-is for now, perhaps introducing a new term if the docs team can settle on one.

Fine; I'll change it back.

Reply:
Right, I've worked with some of those PLs before, I've not encountered the term in a user context. I'm totally on board with using the term in the reference: but it's a different matter entirely to expose the user to it.

Reply:
Yeah, while I think both both of you are right in your own ways here (those languages – mainly quite "technical" ones, it should be noted – may use the term, especially in their language references and/or diagnostics), but @Manishearth makes a sound point when he says "head expression" may be a new compound term, but should have a meaning that is pretty straightforward to guess from its component words. The pattern of forming "-ee" suffixes for object (i.e. passive) nouns and the verb "scrutinise" to begin with are things that many foreign speakers won't know, I suspect. And even for me I didn't immediately guess what was meant by "scrutinee" when I first saw it used (without an example). Being newbie-friendly is generally good. :-)